game/ffmpeg.cc

```diff
@@ -170,85 +159,129 @@ FFmpeg::FFmpeg(fs::path const& _filename, unsigned int rate):
+		av_opt_set_int(m_resampleContext.get(), "out_channel_layout", av_get_default_channel_layout(AUDIO_CHANNELS), 0);
+		av_opt_set_int(m_resampleContext.get(), "in_sample_rate", m_codecContext->sample_rate, 0);
+		av_opt_set_int(m_resampleContext.get(), "out_sample_rate", m_rate, 0);
+		av_opt_set_int(m_resampleContext.get(), "in_sample_fmt", m_codecContext->sample_fmt, 0);
+		av_opt_set_int(m_resampleContext.get(), "out_sample_fmt", AV_SAMPLE_FMT_S16, 0);
+		swr_init(m_resampleContext.get());
 		if (!m_resampleContext) throw std::runtime_error("Cannot create resampling context");
 		audioQueue.setSamplesPerSecond(AUDIO_CHANNELS * m_rate);
 		break;
 	case AVMEDIA_TYPE_VIDEO:
 		// Setup software scaling context for YUV to RGB conversion
 		width = m_codecContext->width;
 		height = m_codecContext->height;
-		m_swsContext = sws_getContext(
+		m_swsContext.reset(sws_getContext(
 		  m_codecContext->width, m_codecContext->height, m_codecContext->pix_fmt,
 		  width, height, AV_PIX_FMT_RGB24,
-		  SWS_POINT, nullptr, nullptr, nullptr);
+		  SWS_POINT, nullptr, nullptr, nullptr));
 		break;
 	default:  // Should never be reached but avoids compile warnings
 		abort();
 	}
 }
 
+class FfmpegError: public std::runtime_error {
+public:
+	FfmpegError(int errorValue): runtime_error(msgFmt(errorValue)) {}
+private:
+	static std::string msgFmt(int errorValue) {
+		char message[AV_ERROR_MAX_STRING_SIZE];
+		av_strerror(errorValue, message, AV_ERROR_MAX_STRING_SIZE);
+		std::ostringstream oss;
+		oss << "FfmpegError: code=" << errorValue << ", error=" << message;
+		return oss.str();
+	}
+};
+
+struct FFmpeg::Packet: public AVPacket {
+	Packet() {
+		av_init_packet(this);
+		this->data = nullptr;
+		this->size = 0;
+	}
+
+	~Packet() { av_packet_unref(this); }
+};
+
 void FFmpeg::operator()() {
 	try { open(); } catch (std::exception const& e) { std::clog << "ffmpeg/error: Failed to open " << m_filename << ": " << e.what() << std::endl; return; }
 	m_duration = m_formatContext->duration / double(AV_TIME_BASE);
 	audioQueue.setDuration(m_duration);
 	int errors = 0;
 	bool eof = false;
 	std::clog << "audio/debug: FFmpeg processing " << m_filename.filename().string() << std::endl;
-	while (!terminating()) {
-		if (eof) break;
+	Packet pkt;
+	while (!terminating() && !eof) {
 		try {
-			if (audioQueue.wantSeek()) m_seekTarget = 0.0;
-			if (m_seekTarget == m_seekTarget) seek_internal();
-			decodePacket();
-			errors = 0;
+			auto ret = av_read_frame(m_formatContext.get(), &pkt);
+			if(ret == AVERROR_EOF) {
+				// End of file: no more data to read.
+				throw FFmpeg::eof_error();
+			} else if(ret < 0) {
+				throw FfmpegError(ret);
+			}
+
+			if (pkt.stream_index == m_streamId) {
+
+				if (audioQueue.wantSeek()) m_seekTarget = 0.0;
+				if (m_seekTarget == m_seekTarget) seek_internal();
+				decodePacket(pkt);
+				errors = 0;
+			}
```

Comment:
Does this (still) support seeking back from near to the end of a song, once the decoder has already reached EOF?

Reply:
@OznOg can you test/check/verify this :)?

Reply:
Running a song almost till the end and then rewind still works!
There is still a bug from a few years ago https://github.com/performous/performous/issues/33
which sadly still applies.